contributors/design-proposals/storage/csi-migration.md

```diff
@@ -237,26 +237,55 @@ with the in-tree plugin, the VolumeAttachment object becomes orphaned.
 ### In-line Volumes
 
 In-line controller calls are a special case because there is no PV. In this case
-we will add the CSI Source JSON to the VolumeToAttach object and in Attach we
-will put the Source in a new field in the VolumeAttachment object
-VolumeAttachment.Spec.Source.VolumeAttachmentSource.InlineVolumeSource. The CSI Attacher will have to
-be modified to also check this location for a source before checking the PV
-itself.
+we will forward the in-tree volume source to CSI attach as-is and it will be
+copied to a new field in the VolumeAttachment object
+VolumeAttachment.Spec.Source.VolumeAttachmentSource.InlineVolumeSource. The
```

Comment:
The `InlineVolumeSource` member introduced is of type [CSIPersistentVolumeSource](https://github.com/davidz627/kubernetes/commit/9dbae27ccbd3e0cacedd567f34ba302fb94ee893#diff-7dc0dea2b32ece49304df77954ffb69aR165). Does this design require an inline volume (of type `Volume`) to be translated to type `PersistentVolume`?

If we create a dummy `PersistentVolume` object and set it's `Spec.PersistentVolumeSource` field to store the details for the inline `Volume` as part of [translate](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/operationexecutor/operation_generator.go#L323) calls before CSI `attach()` is called, then:
1. The translation of a `VolumeSource` to a `PersistentVolumeSource` feels like a hack.
2. CSI `attach()` won't have a clean way to differentiate between inline and PV scenarios based on the `spec` it [receives](https://github.com/kubernetes/kubernetes/blob/3624c74ce880168eb456f15ef50a92d1cbff8da1/pkg/volume/csi/csi_attacher.go#L60) since in both in-line and regular scenarios, `spec.PersistentVolume` will be non null.

A cleaner way may be to translate an inline `Volume`'s source to a `CSIVolumeSource` introduced [recently](https://github.com/kubernetes/kubernetes/commit/4ec7d2305d8c022fe7870d0e8f53c27045e84b96#diff-c5f41a1fd7d92b2773d2d5194a0fb6ffR1698) and set the `Volume` to the `spec.Volume` sent to CSI's `attach()`. This is a more "natural" translation as both entities are inline volume sources (rather than one being a persistent volume source) and CSI can easily distinguish between the inline and non-inline scenarios. It will require `CSIVolumeSource` to have an extra `VolumeHandle` field.

Reply:
The idea was not to create a new PV object, but just fill in the PVSource in the VolumeAttachment object.  It does require translating inline volume source to pv volume source, which should be equivalent for all the drivers we plan to migrate.

Is there a need for the attach to be able to distinguish between inline and PV?

The trouble with CSIVolumeSource is that we don't want to actually support persistent use cases as an inline volume.